src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -298,19 +289,31 @@ TransactionManager::submit_transaction_direct(
 {
   LOG_PREFIX(TransactionManager::submit_transaction_direct);
   SUBTRACET(seastore_t, "start", tref);
-  return trans_intr::make_interruptible(
-    tref.get_handle().enter(write_pipeline.ool_writes)
-  ).then_interruptible([this, FNAME, &tref] {
-    return seastar::do_with(tref.get_delayed_alloc_list(),
-			    [this, FNAME, &tref](auto &delayed_extents) {
-      return epm->delayed_allocate_and_write(tref, delayed_extents
-      ).si_then([this, FNAME, &tref, &delayed_extents] {
-        SUBTRACET(seastore_t, "update delayed extent mappings", tref);
-        return lba_manager->update_mappings(tref, delayed_extents);
-      }).handle_error_interruptible(
-        crimson::ct_error::input_output_error::pass_further(),
-        crimson::ct_error::assert_all("invalid error")
-      );
+  return seastar::do_with(epm->dispatch_delayed_extents(tref),
+			  [this, FNAME, &tref](auto& dispatch_res) {
+    auto fut = seastar::now();
+    if (tref.get_src() == Transaction::src_t::MUTATE) {
+      // other transactions has already reserved space in
+      // EPM::BackgroundProcess::do_background_cycle
+      auto usage = tref.get_projected_usage();
+      SUBTRACET(seastore_t, "reserve usage {}", tref, usage);
+      fut = epm->reserve_projected_usage(usage);
```

Comment:
I'm wondering if it will be cleaner to keep `reserve/release_projected_usage()` (also the preceding `dispatch_delayed_extents()`) in the original `submit_transaction()` for `MUTATE` transactions? Then encapsulate the shared logic into a private `do_submit_transaction()`, and called by the `submit_transaction_direct()` also with `dispatch_delayed_extents()`?

Reply:
Just see your replies.

The updated code looks good.